1\. Write a function that converts number representation (bin<->dec<->hex)

In [103]:
#numbers to be converted
x=18
y=0b00100110
z=0xbe

#convert function
def convert(a):
        a_bin=bin(a) 
        a_hex=hex(a) 
        a_dec=int(str(a),10) 
        return(a_dec,a_bin,a_hex)
    
print(convert(x))
print(convert(y))
print(convert(z))

(18, '0b10010', '0x12')
(38, '0b100110', '0x26')
(190, '0b10111110', '0xbe')


2\. Write a function that converts a 32 bit word into a single precision floating point (i.e. interprets the various bits as sign, mantissa and exponent)

In [104]:
#32 bit word
word='10010011011010010100100100100101'

#convert the word into a float
def conv(x):
    
    sign=(-1)**(int(x[0]))#x[0]=sign,one bit
    
    exp=list(x[1:9])#x[1:9]=exponent, 8 bits
    exp.reverse()
    exponent=0
    for i in range(len(exp)):exponent+=int(exp[i])*2**i

    f=x[9:32] #x[10:32] mantissa,22 bits
    mant=1
    for i in range(len(f)):mant+=int(f[i])*2**(-i-1)
    
    #convert the number
    bias=127
    result=sign*mant*2**(exponent-bias)
    
    return result

conv(word)

-2.944479758484797e-27

3\. Write a program to determine the underflow and overflow limits (within a factor of 2) for python on your computer. 

**Tips**: define two variables inizialized to 1 and halve/double them enough time to exceed the under/over-flow limits  

In [105]:
import sys
from math import inf

double=1.0
half=1.0

#double and halve enough times to reach the limits
while double*2!=inf:double*=2 
while half/2!=0:half/=2
    
print("Limits within a factor two:","\n Max:",double,"\n Min:",half)
print("Max_float:",sys.float_info.max,"\nMin_float:",sys.float_info.min)
print("Check:",0<sys.float_info.min<half,inf>sys.float_info.max>double)

Limits within a factor two: 
 Max: 8.98846567431158e+307 
 Min: 5e-324
Max_float: 1.7976931348623157e+308 
Min_float: 2.2250738585072014e-308
Check: False True


COMMENTS:

The value of "half" is smaller than the value of "sys.float_info.min". 
"half" happens to be a denormalized number, meaning that its representation it's different from the one of a normal 32 precision floating point. This change of representation happens automatically, in order to reach a value even nearer to zero.

4\. Write a program to determine the machine precision

**Tips**: define a new variable by adding a smaller and smaller value (proceeding similarly to prob. 2) to an original variable and check the point where the two are the same 

In [106]:
import sys

x=5
y=0
eps=1

#add smaller number every time
while x!=y:
    y=x+eps
    eps/=2
    
print("Determined machine precision:",eps)
print("System machine precision:",sys.float_info.epsilon)
print("Check:",eps==sys.float_info.epsilon)

Determined machine precision: 2.220446049250313e-16
System machine precision: 2.220446049250313e-16
Check: True


5\. Write a function that takes in input three parameters $a$, $b$ and $c$ and prints out the two solutions to the quadratic equation $ax^2+bx+c=0$ using the standard formula:
$$
x=\frac{-b\pm\sqrt{b^2-4ac}}{2a}
$$

(a) use the program to compute the solution for $a=0.001$, $b=1000$ and $c=0.001$

(b) re-express the standard solution formula by multiplying top and bottom by $-b\mp\sqrt{b^2-4ac}$ and again find the solution for $a=0.001$, $b=1000$ and $c=0.001$. How does it compare with what previously obtained? Why?

(c) write a function that compute the roots of a quadratic equation accurately in all cases

In [107]:
from math import sqrt

a=0.001
b=1000
c=a

#a) Standard formula solutions
def solutions(a,b,c):
    sol1=(-b+(b**2-4*a*c)**(1/2))/(2*a)
    sol2=(-b-(b**2-4*a*c)**(1/2))/(2*a)
    return[sol1,sol2]
print("Standard formula solutions:",solutions(a,b,c))

#b) New formula solutions
def new_solutions(a,b,c):
    sol1=(2*c)/(-b-(b**2 -4*a*c)**(1/2))
    sol2=(2*c)/(-b+(b**2 -4*a*c)**(1/2))
    return[sol1,sol2]
print("New formula solutions:",new_solutions(a,b,c))

#c) Always correct solutions
def quadratic(a,b,c):
    x=b**2
    y=4*a*c
    det=sqrt(x-y)
    den=2*a
    sol1=-b+det
    sol2=-b-det
    return[sol1/den,sol2/den]
print("Always correct solutions:",quadratic(a,b,c))

Standard formula solutions: [-9.999894245993346e-07, -999999.999999]
New formula solutions: [-1.000000000001e-06, -1000010.5755125057]
Always correct solutions: [-9.999894245993346e-07, -999999.999999]


6\. Write a program that implements the function $f(x)=x(x−1)$

(a) Calculate the derivative of the function at the point $x = 1$ using the derivative definition:

$$
\frac{{\rm d}f}{{\rm d}x} = \lim_{\delta\to0} \frac{f(x+\delta)-f(x)}{\delta}
$$

with $\delta = 10^{−2}$. Calculate the true value of the same derivative analytically and compare with the answer your program gives. The two will not agree perfectly. Why not?

(b) Repeat the calculation for $\delta = 10^{−4}, 10^{−6}, 10^{−8}, 10^{−10}, 10^{−12}$ and $10^{−14}$. How does the accuracy scales with $\delta$?

In [68]:
#list of deltas
delta=[10**(-n) for n in range(1,15) if n%2==0]

#define the function
def f(x):return x*(x-1)

#define function that calculates the derivative
def derivative_f(delta):
    num=f(1+delta)-f(1)
    return num/delta

#a)
print("Derivative for delta=0.01:",derivative_f(delta[0]))

#b)
derivatives=[derivative_f(val) for val in delta]
print("Derivatives for all deltas:",derivatives)

accuracy=[val-1 for val in derivatives]
print("Accuracy for all the values:",accuracy)

Derivative for delta=0.01: 1.010000000000001
Derivatives for all deltas: [1.010000000000001, 1.0000999999998899, 1.0000009999177333, 1.0000000039225287, 1.000000082840371, 1.0000889005833413, 0.9992007221626509]
Accuracy for all the values: [0.010000000000000897, 9.999999988985486e-05, 9.99917733279787e-07, 3.922528746258536e-09, 8.284037100736441e-08, 8.890058334132256e-05, -0.0007992778373491216]


COMMENTS:

a)
Analytically the derivative of the function in x=1 is equal to one. In the first place it can't be equal to python's result beacause, of course, of the finite non-zero value of delta. Secondarily, the result doesn't come out perfectly 1.01 because python can't store exactely 0.01 as a value of delta; the error is eventually propagated during the calculations concerning delta.

b)The best accuracy is found at $\delta=10^{-8}$, due to the uncapability of python to perform the division $\frac{f(1+\delta)}{\delta}$, with $\delta$ every time ideally nearer to $0$.

7\. Consider the integral of the semicircle of radius 1:
$$
I=\int_{-1}^{1} \sqrt(1-x^2) {\rm d}x
$$
which it's known to be $I=\frac{\pi}{2}=1.57079632679...$.
Alternatively we can use the Riemann definition of the integral:
$$
I=\lim_{N\to\infty} \sum_{k=1}^{N} h y_k 
$$

with $h=2/N$ the width of each of the $N$ slices the domain is divided into and where
$y_k$ is the value of the function at the $k-$th slice.

(a) Write a programe to compute the integral with $N=100$. How does the result compares to the true value?

(b) How much can $N$ be increased if the computation needs to be run in less than a second? What is the gain in running it for 1 minute? 


In [102]:
from math import sqrt,pi

I=pi/2
f=lambda x: sqrt(1-x**2) #integrand function
 
#function that calculates the integral    
def integral(N):
    h=2/N
    x=[-1+i*h for i in range(N+1)]
    integral=0
    for i in range(len(x)):integral+=(h*f(x[i]))
    return integral

#a) N=100
N=100
print("Calculated integral:",integral(N),"\nExpected value:",I)
print("Error of the measure with N=100 relative to I: %.2f" % ((I-integral(N))*100/I),"%")


#b)
import timeit

N=[10**n for n in range(2,7)] #list of Ns

#print and time the integral for every N
for i in range(len(N)):
    print("Result of integral:%.f/%f" % (integral(N[i]),I))
    print("Relative error of measure with N=%d:"%N[i],(I-integral(N[i]))*100/I,"%.")
    %timeit integral(N[i])

print("Computation in less than a second, max_N=",N[4])
print("Computation in less than a minute, max_N=",100*N[4])

Calculated integral: 1.5691342555492505 
Expected value: 1.5707963267948966
Error of the measure with N=100 relative to I: 0.11 %
Result of integral:2/1.570796
Relative error of measure with N=100: 0.10581074180619238 %.
32.8 µs ± 474 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
Result of integral:2/1.570796
Relative error of measure with N=1000: 0.0033478747644449802 %.
341 µs ± 6.13 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
Result of integral:2/1.570796
Relative error of measure with N=10000: 0.00010587493599366788 %.
3.44 ms ± 60.9 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
Result of integral:2/1.570796
Relative error of measure with N=100000: 3.3480772388811256e-06 %.
34.7 ms ± 446 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
Result of integral:2/1.570796
Relative error of measure with N=1000000: 1.058806370743233e-07 %.
372 ms ± 9.91 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
Computation in less than a second, max_N

COMMENTS:

b)
Calculating the integral with different values of N, I have noticed that increasing N by a factor 10, the run time of the function **integral()** is increased by a factor $10$. Python allows me to run codes only up to $1$ sec of run time, so I have inferred that:

-if the code needs to be run in less than $1$ sec, then N can reach the maximum value of about $10^{6}$, leading to an accuracy of measurement of $1.05\times10^{-07}\%$

-if the code needs to be run in less than $1$ min, then N can reach the maximum value of about $10^{9}$, leading to an accuracy of measurement of $1.05\times10^{-09}\%$